reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,423 @@
+                                )
+                                resources.append(ers)
+
+                break
+
+        return resources
+
+    def resource_needs_reconciliation(
+        self,
+        resource: ExternalResource,
+        state: ExternalResourceState,
+    ) -> bool:
+        if state.status == ResourceStatus.NOT_EXISTS:
+            logging.info("External resource does not exist -> Needs Reconcile")
+            return True
+        # elif (
+        #     state.status == ResourceStatus.IN_PROGRESS
+        #     and (datetime.now(state.ts.tzinfo) - state.ts).total_seconds() > 3600
+        # ):
+        #     logging.info(
+        #         "External Resources Reconcile-TTL reached -> Needs Reconcile ??"
+        #     )
+        #     return True
+        # elif state.status == ResourceStatus.ERROR:
+        #     logging.info("External Resources In ERROR state -> ??")
+        #     return True
+        elif state.status == ResourceStatus.CREATED:
+            if (datetime.now(state.ts.tzinfo) - state.ts).days > 0:
+                logging.info(
+                    "External Resources periodic drift reconciliation run -> Needs Reconcile"
+                )
+                return True
+            elif state.resource_hash != resource.digest():
+                logging.info("Resource attributes has changed -> Needs Reconcile")
+                return True
+
+        logging.info(
+            "External resource does not need reconcile. Status=" + state.status
+        )
+        return False
+
+    def resource_can_be_deleted(self, state: ExternalResourceState) -> bool:
+        if state.status == ResourceStatus.CREATED:
+            return True
+        return False
+
+    def handle_desired_resources(
+        self,
+        specs: Iterable[ExternalResourceSpec],
+    ) -> None:
+        for spec in specs:
+            key = ExternalResourceKey.from_spec(spec)
+            try:
+                state = self.state_mgr.get_external_resource_state(key)
+                if state.status == ResourceStatus.IN_PROGRESS:
+                    reconcile_status = self.impl.get_resource_reconcile_status(key)
+                    if reconcile_status != ReconcileStatus.IN_PROGRESS:
+                        state.ts = datetime.now(timezone.utc)
+                        if reconcile_status == ReconcileStatus.SUCCESS:
+                            state.status = ResourceStatus.CREATED
+                        elif reconcile_status == ReconcileStatus.ERROR:
+                            state.status = ResourceStatus.ERROR
+                        logging.info("Resource Status updated to " + state.status)
+            except StateNotFoundError:
+                state = ExternalResourceState(
+                    key=key,
+                    ts=datetime.now(timezone.utc),
+                    status=ResourceStatus.NOT_EXISTS,
+                )
+            except Exception:
+                raise
```

Comment:
try block should be limit to only related to expected error `StateNotFoundError`, it looks like only `self.state_mgr.get_external_resource_state(key)` can throw it, but I can't find the location throwing it. And if no handling for generic `Exception`, no need to except it.

Worth extract method to simplify it:

```python
    def _get_state(self, key: ExternalResourceKey) -> ExternalResourceState:
        try:
            state = self.state_mgr.get_external_resource_state(key)
        except StateNotFoundError:
            return ExternalResourceState(
                key=key,
                ts=datetime.now(timezone.utc),
                status=ResourceStatus.NOT_EXISTS,
            )

        if state.status != ResourceStatus.IN_PROGRESS:
            return state

        match self.impl.get_resource_reconcile_status(key):
            case ReconcileStatus.SUCCESS:
                state.status = ResourceStatus.CREATED
            case ReconcileStatus.ERROR:
                state.status = ResourceStatus.ERROR
            case _:
                return state

        state.ts = datetime.now(timezone.utc)
        logging.info("Resource Status updated to " + state.status)
        return state
```

Reply:
This also changed. I close this comment, PTAL to the new code.